### PR TITLE
Fix #2671: Interrupt R is disabled but REPL is busy

### LINKED ISSUE #2671: Interrupt R is disabled but REPL is busy
RTVS_2016-11-08.3

I've issued some plot commands to remote, and it was stuck, so I used Interrupt R.  I did that twice.

Then I tried `1+1` which also hanged, but Interrupt R is disabled.

image

Dump in \\pytools\r\Bugs\2671

Things work fine again after clicking Reset.

### PR DESCRIPTION
**Pull request by int19h**
When interrupting R, do not cancel the current interaction request task until R confirms completion.

Disable "Interrupt R" command while it's running to prevent re-entrancy.

Fix interrupt test to wait for prompt before first checking command status.

### CODE REVIEW

File: src/Host/Client/Impl/Session/RSession.cs [hunk 1]
```diff
@@ -203,14 +203,13 @@ internal sealed class RSession : IRSession, IRCallbacks {
...
-                var cancelTask = _host.CancelAllAsync(cancellationToken);
+                await _host.CancelAllAsync(cancellationToken);
 
                 var currentRequest = Interlocked.Exchange(ref _currentRequestSource, null);
                 var exception = new OperationCanceledException();
                 currentRequest?.TryCancel(exception);
-                ClearPendingRequests(exception);
 
-                await cancelTask;
+                ClearPendingRequests(exception);
```
Comment (AlexanderSher) [inline src/Host/Client/Impl/Session/RSession.cs]: Do you really need to call ` ClearPendingRequests` after `CancelAllAsync`? I thought it was about `currentRequest?.TryCancel(exception)`

Comment (int19h) [inline src/Host/Client/Impl/Session/RSession.cs]: This is not new code - that line was there already.

File: src/R/Components/Test/InteractiveWorkflow/RInteractiveWorkflowCommandTest.cs [hunk 2]
```diff
@@ -106,6 +106,7 @@ public class RInteractiveWorkflowCommandTest : IDisposable {
             command.Should().BeInvisibleAndDisabled();
 
             using (await UIThreadHelper.Instance.Invoke(() => _workflow.GetOrCreateVisualComponentAsync())) {
+                using (await _workflow.RSession.BeginInteractionAsync()) { }
```
Comment (AlexanderSher) [inline src/R/Components/Test/InteractiveWorkflow/RInteractiveWorkflowCommandTest.cs]: I've seen this issue too.
Not part of this change, but maybe we should set `RSession._initializationTcs` not in `IRCallbacks.Connected`, but in `IRCallbacks.ReadConsole`?

File: src/Host/Client/Impl/Session/RSession.cs (hunk 1 above)
Comment (AlexanderSher) [inline src/Host/Client/Impl/Session/RSession.cs]: It was, but previously it was before `await cancelTask;`

File: src/R/Components/Test/InteractiveWorkflow/RInteractiveWorkflowCommandTest.cs (hunk 2 above)
Comment (int19h) [inline src/R/Components/Test/InteractiveWorkflow/RInteractiveWorkflowCommandTest.cs]: These sorts of issues were part of my motivation suggesting that `Connected` (as in, the event that's exposed outside of `RSession`) should perhaps only be fired on the first prompt, and everything before that prompt should be treated as internal initialization. What you suggest would be one way to do just that, if I understand correctly :)

File: src/Host/Client/Impl/Session/RSession.cs (hunk 1 above)
Comment (int19h) [inline src/Host/Client/Impl/Session/RSession.cs]: Indeed. Will do.

File: src/R/Components/Test/InteractiveWorkflow/RInteractiveWorkflowCommandTest.cs (hunk 2 above)
Comment (int19h) [inline src/R/Components/Test/InteractiveWorkflow/RInteractiveWorkflowCommandTest.cs]: Either way, I'm all for it.

**Comment by AlexanderSher** [comment]
👍
